File: 0079-word-search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rowmax, colmax = len(board), len(board[0])
        def backtrack(r, c, amountright):
            if amountright == len(word):
                return True
            if r < 0 or c < 0 or r >= rowmax or c >= colmax or board[r][c] != word[amountright]:
                return False
            
            temp = board[r][c]
            board[r][c] = "#"
            found = (backtrack(r, c - 1, amountright + 1) or 
                    backtrack(r, c + 1, amountright + 1) or 
                    backtrack(r + 1, c, amountright + 1) or 
                    backtrack(r - 1, c, amountright + 1))
            board[r][c] = temp
            return found

        
        for r in range(rowmax):
            for c in range(colmax):
                if board[r][c] == word[0] and backtrack(r, c, 0):
                    return True
        return False

File: 0079-word-search/solution.cpp
class Solution {
public:
    int ROWS;
    int COLS;
    bool backtrack(int r, int c, int wordindex, vector<vector<char>>& board, const string& word){
        if (wordindex == word.size()){
            return true;    
        }
        if(r < 0 or c < 0 or ROWS <= r or COLS <= c or board[r][c] != word[wordindex]){
            return false;
        }
        char temp = board[r][c];
        board[r][c] = '#';
        if(backtrack(r + 1, c, wordindex + 1, board, word) or backtrack(r, c + 1, wordindex + 1, board, word) or backtrack(r - 1, c, wordindex + 1, board, word) or backtrack(r, c - 1, wordindex + 1, board, word)){
            return true;
        }
        board[r][c] = temp;
        return false;
    }
    bool exist(vector<vector<char>>& board, string word) {
        ROWS = board.size();
        COLS = board[0].size();
        string path = "";
        for(int r = 0; r < ROWS; ++r){
            for (int c = 0; c < COLS; ++c){
                if(board[r][c] == word[0] and backtrack(r, c, 0, board, word)){
                    return true;
                }
            }
        }
        return false;
    }
};
